src/org/exist/xquery/functions/fn/FunId.java

```diff
@@ -217,13 +193,15 @@ private void getId(NodeSet result, DocumentSet docs, String id) throws XPathExce
 
     private void getId(Sequence result, Sequence seq, String id) throws XPathException {
         final Set<DocumentImpl> visitedDocs = new TreeSet<>();
-        for (final SequenceIterator i = seq.iterate(); i.hasNext();) {
+        for (final SequenceIterator i = seq.iterate(); i.hasNext(); ) {
             final NodeImpl v = (NodeImpl) i.nextItem();
             final DocumentImpl doc = v.getOwnerDocument();
-            if (!visitedDocs.contains(doc)) {
+
+            if (doc != null && !visitedDocs.contains(doc)) {
```

Comment:
So the problem is that `org.exist.dom.memtree.NodeImpl#getOwnerDocument()` returns `null`? If I understand correctly, this will only happen when the `NodeImpl` is a `org.exist.dom.memtree.DocumentImpl`?

If that is the case, then this fix looks fine to me. I would have liked to have seen a test though ;-)

Reply:
yes, in a commit from your hand, `org.exist.dom.memtree.NodeImpl#getOwnerDocument()` can return null.

Reply:
I agree with that we need a test......

Reply:
I studied your commit, where the `org.exist.dom.memtree.NodeImpl#getOwnerDocument()` can actually return null ; Being that a valid value, this loop should be resistent for it.